sessionctx/variable/session.go

```diff
@@ -316,6 +333,62 @@ func (tc *TransactionContext) GetStmtFutureForRC() oracle.Future {
 	return tc.stmtFuture
 }
 
+// AddSavepoint adds a new savepoint.
+func (tc *TransactionContext) AddSavepoint(name string, savepoint interface{}) {
+	name = strings.ToLower(name)
+	tc.DeleteSavepoint(name)
+
+	tableDeltaMap := make(map[int64]TableDelta, len(tc.TableDeltaMap))
+	for k, v := range tc.TableDeltaMap {
+		tableDeltaMap[k] = v.Clone()
+	}
+	pessimisticLockCache := make(map[string][]byte, len(tc.pessimisticLockCache))
+	for k, v := range tc.pessimisticLockCache {
+		pessimisticLockCache[k] = v
+	}
```

Comment:
Better to abstract and wrap a standalone function or interface to reset/restore the transaction status so that nothing would be missing.

Reply:
Already add `GetCurrentSavepoint` and `RestoreBySavepoint` method for `TransactionContext`, PTAL.